plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSource.java

```diff
@@ -132,140 +150,83 @@ public long getReadTimeNanos()
+                    else if (longReadFunctions[i] != null) {
+                        type.writeLong(output, longReadFunctions[i].readLong(resultSet, i + 1));
+                    }
+                    else if (sliceReadFunctions[i] != null) {
+                        type.writeSlice(output, sliceReadFunctions[i].readSlice(resultSet, i + 1));
+                    }
+                    else {
+                        type.writeObject(output, objectReadFunctions[i].readObject(resultSet, i + 1));
                     }
-                    throw new RuntimeException(e);
-                }
-                catch (InterruptedException e) {
-                    Thread.currentThread().interrupt();
-                    resultSetFuture.cancel(true);
-                    throw new RuntimeException(e);
-                }
-                finally {
-                    readTimeNanos.addAndGet(System.nanoTime() - start);
                 }
             }
-            return resultSet.next();
-        }
-        catch (SQLException | RuntimeException e) {
-            throw handleSqlException(e);
-        }
-    }
 
-    @Override
-    public boolean getBoolean(int field)
-    {
-        checkState(!closed, "cursor is closed");
-        requireNonNull(resultSet, "resultSet is null");
-        try {
-            return booleanReadFunctions[field].readBoolean(resultSet, field + 1);
+            if (!pageBuilder.isFull()) {
+                finished = true;
+            }
         }
         catch (SQLException | RuntimeException e) {
             throw handleSqlException(e);
         }
-    }
 
-    @Override
-    public long getLong(int field)
-    {
-        checkState(!closed, "cursor is closed");
-        requireNonNull(resultSet, "resultSet is null");
-        try {
-            return longReadFunctions[field].readLong(resultSet, field + 1);
-        }
-        catch (SQLException | RuntimeException e) {
-            throw handleSqlException(e);
-        }
+        Page page = pageBuilder.build();
+        pageBuilder.reset();
+        return page;
     }
 
     @Override
-    public double getDouble(int field)
+    public long getMemoryUsage()
     {
-        checkState(!closed, "cursor is closed");
-        requireNonNull(resultSet, "resultSet is null");
-        try {
-            return doubleReadFunctions[field].readDouble(resultSet, field + 1);
-        }
-        catch (SQLException | RuntimeException e) {
-            throw handleSqlException(e);
-        }
+        return pageBuilder.getRetainedSizeInBytes();
     }
 
     @Override
-    public Slice getSlice(int field)
+    public long getCompletedBytes()
     {
-        checkState(!closed, "cursor is closed");
-        requireNonNull(resultSet, "resultSet is null");
-        try {
-            return sliceReadFunctions[field].readSlice(resultSet, field + 1);
-        }
-        catch (SQLException | RuntimeException e) {
-            throw handleSqlException(e);
-        }
+        return 0;
```

Comment:
we can calculate this, right?

Reply:
Can we repot size of the pages created? This is a rough estimate on the size of the data since we are not using dictionary encoding here. Most of the databases are using binary framing which is similiar to our in memory structures.

Reply:
Processed input bytes is already tracking size of pages and that is done by the engine independently of the connector, this field is about the connector reporting physical input bytes

Reply:
Sure, ship it!